### PR TITLE
New unit conversion functionality

### PR DESCRIPTION
**Pull request by mikejmets**
Created because #526 what created on the old repo.

## Description of the issue/feature this PR addresses
This new feature allows analysis result value to be reported in a unit other than one in which the result was recorded. (eg. analysis value is entered as mg/l but COA requires mg/g). This PR includes:
* a new setup data `UnitConversion` content type with a unit and a formula
* a new DataGridField on `AnalysisService` for `UnitConversions`.  The DGF includes `SampleType` to allow different conversions for each sample type and 2 booleans, `HidePrimaryUnit` to allow the primary value to be excluded from the COA and `ShowOnListing` to allow a converted result to appear on the COA but not the analyses listing.
* a doctest

--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

### CODE REVIEW

File: bika/lims/browser/analysisrequest/publish.py [hunk 1]
```diff
@@ -726,7 +729,7 @@ def __call__(self, ar, overwrite=False):
         verified = wasTransitionPerformed(ar, 'verify')
         if not overwrite and verified:
             # Prevent any error related with digest
-            data = ar.getDigest() if hasattr(ar, 'getDigest') else {}
+            data = {} #ar.getDigest() if hasattr(ar, 'getDigest') else {}
```
Comment (ramonski) [inline bika/lims/browser/analysisrequest/publish.py]: Although I also want to remove the Digester, we should do that in a separate PR;)

File: bika/lims/browser/analysisrequest/publish.py [hunk 2]
```diff
@@ -1312,10 +1316,29 @@ def _analyses_data(self, ar, analysis_states=None):
                 andict['previous_results'] = ", ".join(
                     [p['formatted_result'] for p in andict['previous'][-5:]])
 
-            analyses.append(andict)
+            #Append addition analysis for each unit conversion
+            if andict['unit_conversions']:
```
Comment (ramonski) [inline bika/lims/browser/analysisrequest/publish.py]: This looks like it might fail with a `KeyError`

File: bika/lims/browser/analysisrequest/publish.py [hunk 3]
```diff
@@ -1312,10 +1316,29 @@ def _analyses_data(self, ar, analysis_states=None):
...
+                    if andict.get('result'):
+                        new['formatted_result'] = new['result'] = \
+                            convert_unit(
+                                    andict['result'], 
+                                    unit_conversion.formula,
+                                    andict['obj'].getPrecision())
+                    analyses.append(new)
+            #Append original analysis
+            if not andict.get('hide_primary_result'):
+                analyses.append(andict)
```
Comment (ramonski) [inline bika/lims/browser/analysisrequest/publish.py]: please refactor that to a more maintainable function, e.g.:

```python
def convert_unit_of(uid):
    obj = api.get_object_by_uid(uid)
    return {
        "unit": obj. converted_unit,
        "formatted_unit": obj.converted_unit,
         ...
    }
```

Then you can just do something like this:

```python
uc = convert_unit_of(uc_uid)
if uc:
    an_dict.update(uc)
```

File: bika/lims/browser/analysisrequest/publish.py [hunk 4]
```diff
@@ -1345,7 +1368,10 @@ def _analysis_data(self, analysis, decimalmark=None):
                       else None,
                   'worksheet': None,
                   'specs': {},
-                  'formatted_specs': ''}
+                  'formatted_specs': '',
+                  'unit_conversions': [],
+                  'hide_primary_result': False,
```
Comment (ramonski) [inline bika/lims/browser/analysisrequest/publish.py]: better inject (dict.update) that with your own function and do not rely on the key in another function

File: bika/lims/browser/analysisrequest/publish.py [hunk 5]
```diff
@@ -1404,6 +1430,22 @@ def _analysis_data(self, analysis, decimalmark=None):
...
+            st_uid = sample['sample_type']['obj'].UID()
+        if st_uid:
+            for unit_conversion in analysis.getUnitConversions():
+                if unit_conversion.get('SampleType') and \
+                   unit_conversion.get('Unit') and \
+                   unit_conversion.get('SampleType') == st_uid:
+                    andict['unit_conversions'].append(unit_conversion['Unit'])
+                    if andict['hide_primary_result'] == False:
+                        andict['hide_primary_result'] = \
+                            unit_conversion.get('HidePrimaryUnit', '0') == '1'
```
Comment (ramonski) [inline bika/lims/browser/analysisrequest/publish.py]: Please refactor that, it is not understandable and it looks like a huge regular expression. Sorry for being harsh, but this is not maintainable.

File: bika/lims/browser/calcs.py [hunk 6]
```diff
@@ -93,6 +95,9 @@ def calculate(self, uid=None):
             if form_result == "0/0":
                 Result['result'] = ""
 
+        if Result['result']:
+            Result['converted_result'] = resolve_unit(analysis, Result['result'])
```
Comment (ramonski) [inline bika/lims/browser/calcs.py]: Please refactor to something like this:

```python
result = None
try:
    result = float(form_result)
except:
    result = ""
if result:
    result = resolve_unit(analysis, result)
Result = {"uid": uid, "result": result}
```

File: bika/lims/content/analysisservice.py [hunk 7]
```diff
@@ -602,5 +601,4 @@ def after_deactivate_transition_event(self):
             template = api.get_object(template)
             template.remove_service(self)
 
-
 registerType(AnalysisService, PROJECTNAME)
```
Comment (ramonski) [inline bika/lims/content/analysisservice.py]: the blank lines can be left here, they are according to PEP8

File: bika/lims/content/calculation.py [hunk 8]
```diff
@@ -231,7 +231,7 @@ def getCalculationDependencies(self, flat=False, deps=None):
                 deps[service.UID()] = {}
         return deps
 
-    def getCalculationDependants(self):
+    def getCalculationDependants(self, deps=[]):
```
Comment (ramonski) [inline bika/lims/content/calculation.py]: careful, this needs to get changed.

File: bika/lims/content/unitconversion.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        )
+
+
+
+@indexer(IUnitConversion)
+def title(obj):
+    if obj.title:
+        return obj.title
+
+
```
Comment (ramonski) [inline bika/lims/content/unitconversion.py]: please use consistent amount of blank lines between your functions.

File: bika/lims/controlpanel/bika_unitconversions.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+
+class UnitConversionsView(BikaListingView):
+    """Displays all system's unit conversions
+    """
+    implements(IFolderContentsView, IViewView)
+
+    def __init__(self, context, request):
+        super(UnitConversionsView, self).__init__(context, request)
+        self.catalog = "portal_catalog"
```
Comment (ramonski) [inline bika/lims/controlpanel/bika_unitconversions.py]: shouldn't that be `bika_setup_catalog`?

File: bika/lims/tests/doctests/UnitConversions.rst [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    <SampleType at /plone/bika_setup/bika_sampletypes/sampletype-1>
+
+A `AnalysisService` lives in `/bika_setup/bika_analysisservices` folder.::
+
+    >>> services = self.portal.bika_setup.bika_analysisservices
+    >>> aserv = api.create(services, 'AnalysisService', title='Ca')
+    >>> aserv.setUnitConversions([{'SampleType': stype.UID(), 'unit': unitconv.UID()},])
+    >>> aserv
+    <AnalysisService at /plone/bika_setup/bika_analysisservices/analysisservice-1>
+
```
Comment (ramonski) [inline bika/lims/tests/doctests/UnitConversions.rst]: Please show in this test how the actual unit conversion works with the provided formula

File: bika/lims/utils/__init__.py [hunk 12]
```diff
@@ -890,6 +891,54 @@ def is_bika_installed():
     qi = api.portal.get_tool("portal_quickinstaller")
     return qi.isProductInstalled("bika.lims")
 
+def convert_unit(result, formula, precision):
+    """ take a value and a unit conversion formula and convert the result
+    """
+    try:
+        formula = formula.replace('Value', '%f')
+        new =  eval(formula % float(result))
```
Comment (ramonski) [inline bika/lims/utils/__init__.py]: I would reduce here the set of __builtins__ (see Calculation module) to make the eval not so evil as it might be

File: bika/lims/utils/__init__.py [hunk 13]
```diff
@@ -890,6 +891,54 @@ def is_bika_installed():
...
+        fmt = '{{:.{}f}}'.format(precision)
```
Comment (ramonski) [inline bika/lims/utils/__init__.py]: This format always uses `.` as decimal mark, which might not fit for all users. Better take the decimal mark as provided in the setup

**Review by ramonski — CHANGES_REQUESTED** [review]
Hi Mike, there are some things that need to get changed before we can accept it in the core. Sorry for being here and there harsh with my criticism, I was just focussing on the code;)

File: bika/lims/api.py [hunk 14]
```diff
@@ -464,11 +461,11 @@ def get_object_by_uid(uid, default=_marker):
...
     # we defined the portal object UID to be '0'::
     if uid == '0':
-        return get_portal()
+        return get_portal(context=context)
 
     # we try to find the object with both catalogs
-    pc = get_portal_catalog()
-    uc = get_tool("uid_catalog")
+    pc = get_portal_catalog(context=context)
+    uc = get_tool("uid_catalog", context=context)
```
[the remark this answers is not in the document]
Comment (mikejmets) [reply bika/lims/api.py]: I tried various ways to localise changes to get_tool but none worked. This is the error I get http://docs.plone.org/develop/plone.api/docs/api/exceptions.html#plone.api.exc.CannotGetPortalError

File: bika/lims/api.py [hunk 15]
```diff
@@ -146,28 +146,25 @@ def create(container, portal_type, *args, **kwargs):
     types_tool = get_tool("portal_types")
     fti = types_tool.getTypeInfo(portal_type)
 
-    if fti.product:
+    if fti.product: 
+        # AT content type
         obj = _createObjectByType(portal_type, container, tmp_id)
+        obj.edit(**kwargs)
```
[the remark this answers is not in the document]
Comment (mikejmets) [reply bika/lims/api.py]: Can't remember the error now but it threw an error.

File: bika/lims/api.py [hunk 16]
```diff
@@ -673,12 +670,12 @@ def safe_getattr(brain_or_object, attr, default=_marker):
             attr, repr(brain_or_object)))
 
 
-def get_portal_catalog():
+def get_portal_catalog(context=None):
```
[the remark this answers is not in the document]
Comment (mikejmets) [reply bika/lims/api.py]: Same issue as above

File: bika/lims/api.py [hunk 17]
```diff
@@ -673,12 +670,12 @@ def safe_getattr(brain_or_object, attr, default=_marker):
...
     """Get the portal catalog tool
 
     :returns: Portal Catalog Tool
     """
-    return get_tool("portal_catalog")
+    return get_tool("portal_catalog", context=context)
```
[the remark this answers is not in the document]
Comment (mikejmets) [reply bika/lims/api.py]: Same issue as above

File: bika/lims/browser/analyses.py [hunk 18]
```diff
@@ -1147,3 +1164,4 @@ def folderitems(self):
         # Sort items
         items = sorted(items, key=itemgetter('sortcode'))
         return items
+
```
[the remark this answers is not in the document]
Comment (mikejmets) [reply bika/lims/browser/analyses.py]: Done

File: bika/lims/browser/analysisrequest/publish.py (hunk 1 above)
Comment (mikejmets) [reply bika/lims/browser/analysisrequest/publish.py]: If I uncomment this my code won't work. Please advise.

File: bika/lims/browser/analysisrequest/publish.py (hunk 2 above)
Comment (mikejmets) [reply bika/lims/browser/analysisrequest/publish.py]: Changed to a get statement

File: bika/lims/api.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (ramonski) [reply bika/lims/api.py]: Ok, let's try if we can handle that somehow in `get_tool`/`get_portal` only, so that we do not have to change the signature of all other functions.

Please add this to the `api.get_tool`:

```python
def get_tool(name, context=None, default=_marker):
    # Try first with the context
    if context is not None:
        try:
            context = get_object(context)
            return getToolByName(context, name)
        except (BikaLIMSError, AttributeError) as e:
            # https://github.com/senaite/bika.lims/issues/396
            logger.warn("get_tool::getToolByName({}, '{}') failed: {} "
                        "-> falling back to plone.api.portal.get_tool('{}')"
                        .format(repr(context), name, repr(e), name))
            return get_tool(name, default=default)

    # Try with the plone api
    try:
        return ploneapi.portal.get_tool(name)
    except InvalidParameterError:
        if default is not _marker:
            return default
        fail("No tool named '%s' found." % name)

    # Handle weird Plone API Error
    # https://docs.plone.org/develop/plone.api/docs/api/exceptions.html#plone.api.exc.CannotGetPortalError
    except CannotGetPortalError:
        if default is not _marker:
            return default
        return get_tool(name, context=get_portal(), default=None)
```

Then we have to ensure that `get_portal` does not fail:

```python
def get_app():
        import Zope2
        return Zope2.app()

def get_portal():
    """Get the portal object
    :returns: Portal object
    """
    try
        return ploneapi.portal.getSite()
    except CannotGetPortalError:
        app = get_app()
        sites = app.objectValues("Plone Site")
        if len(sites) != 1:
            return None
        return sites[0]
```

[the remark this answers is not in the document]
Comment (ramonski) [reply bika/lims/api.py]: This is just something that went into my mind, please feel free to optimize and change. If we go that way I would suggest a new PR for the API including an updated doctest.
Name of the PR should be then "Handle CannotGetPortalError in API" or so.

File: bika/lims/browser/analysisrequest/publish.py (hunk 1 above)
Comment (ramonski) [reply bika/lims/browser/analysisrequest/publish.py]: @xispa: The digester seems to cause too many problems, I would suggest that we remove that code for the core as well. IMO the digester is also responsible for having Analyses w/o acquisition chain and I could imagine that the issue regarding the missing site is also because of this. Caching persistent objects is not a good idea.

https://github.com/senaite/senaite.core/blob/master/bika/lims/browser/analysisrequest/publish.py#L915

File: bika/lims/browser/analysisrequest/publish.py (hunk 3 above)
Comment (mikejmets) [reply bika/lims/browser/analysisrequest/publish.py]: I refactored to make it more readable but not they way you did it. Maybe if you see what I did it will be clearer what is going on.

File: bika/lims/browser/analysisrequest/publish.py (hunk 4 above)
Comment (mikejmets) [reply bika/lims/browser/analysisrequest/publish.py]: I am extending the analysis dict with 2 new keys - this is the new dict structure. Are you saying I should initialise it in two steps?

File: bika/lims/api.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (ramonski) [reply bika/lims/api.py]: Maybe the `get_portal` function can be refined with the request, e.g.:
```python
def get_portal():
    """Get the portal object
    :returns: Portal object
    """
    try
        return ploneapi.portal.getSite()
    except CannotGetPortalError:
        app = get_app()

        request = get_request()
        portal_id = "Try to get the portal id out of the request... "
        if portal_id in app.objectIds():
            return app[portal_id]
        sites = app.objectValues("Plone Site")
        if len(sites) != 1:
            return None
        return sites[0]
```

File: bika/lims/browser/analysisrequest/publish.py (hunk 3 above)
Comment (ramonski) [reply bika/lims/browser/analysisrequest/publish.py]: Perfect, I'll review later :+1:
Of course, the purpose of my code snippets is just to give an impulse to show another way and just the first thing that comes into my mind. So it is absolute valid to discard it in favor of your own solution;)

File: bika/lims/browser/analysisrequest/publish.py (hunk 4 above)
Comment (ramonski) [reply bika/lims/browser/analysisrequest/publish.py]: I would just like to avoid something like `andict['unit_conversions'].append(unit_conversion['Unit'])`, because then the reviewer and any following coder has to do the following:

- know that `andict` is a dict (easy)
- imply that the value of the key`'unit_conversion'` is a list (also easy)
- find where this key is added and check that it is always set correctly to avoid a KeyError (getting painful)
-  check where this `unit_conversion` comes from and check what is behind the `Unit` key (painful)
- find all the pieces in the dict that are relevant for the unit conversion to make a small change ...

So try to avoid explicit dictionary operation whenever possible and better make a function that returns the logical parts for a unit conversion in its own dictionary.

File: bika/lims/browser/analysisrequest/publish.py (hunk 5 above)
Comment (mikejmets) [reply bika/lims/browser/analysisrequest/publish.py]: Refactored and added comments

File: bika/lims/browser/calcs.py (hunk 6 above)
Comment (mikejmets) [reply bika/lims/browser/calcs.py]: Done

File: bika/lims/content/analysisservice.py (hunk 7 above)
Comment (mikejmets) [reply bika/lims/content/analysisservice.py]: Done

File: bika/lims/content/unitconversion.py (hunk 9 above)
Comment (mikejmets) [reply bika/lims/content/unitconversion.py]: Done

File: bika/lims/controlpanel/bika_unitconversions.py (hunk 10 above)
Comment (mikejmets) [reply bika/lims/controlpanel/bika_unitconversions.py]: Fixed

File: bika/lims/tests/doctests/UnitConversions.rst (hunk 11 above)
Comment (mikejmets) [reply bika/lims/tests/doctests/UnitConversions.rst]: Completed test

File: bika/lims/utils/__init__.py (hunk 12 above)
Comment (mikejmets) [reply bika/lims/utils/__init__.py]: I made an attempt @ramonski , is it now correct

File: bika/lims/utils/__init__.py (hunk 13 above)
Comment (mikejmets) [reply bika/lims/utils/__init__.py]: Done

File: bika/lims/browser/analysisrequest/publish.py (hunk 1 above)
Comment (xispa) [reply bika/lims/browser/analysisrequest/publish.py]: Agree the digester is causing too many problems, and I am ok to eventually remove it. But until we don't have a new-better machinery for reporting, we need to find a workaround to boost the generation of reports as a replacement of the digester. Some of our clients select ~50 or more ARs for report generation at once, so if we remove the digester, the instance gets freezed (or generation takes too long), which imo, is worst. So, if you want to leave this comment, leave it, but don't remove anything from the digester so meanwhile we can at least keep this single line uncommented for our clients. Another option is to add a key-value in the registry and check it's value here.

File: bika/lims/api.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (mikejmets) [reply bika/lims/api.py]: Also dexterity factory method takes kwargs

File: bika/lims/api.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (mikejmets) [reply bika/lims/api.py]: I tested the above possible solutions but to no avail. The error occurs when publishing in _analyses_data which is called from the digester. It fails in my new unit conversion code but at the start of _analyses_data method are some getToolByName calls that also fail when I change them to get_tool. I'm not saying my solution is pretty but it's not the worst hack ever.

File: bika/lims/browser/analysisrequest/publish.py (hunk 1 above)
Comment (mikejmets) [reply bika/lims/browser/analysisrequest/publish.py]: Uncommented and it now works. Not idea why. I'll continue to watch on the test server

File: bika/lims/api.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (ramonski) [reply bika/lims/api.py]: Not at all, please don't understand it as such.
However, this solution fixes the symptoms, but not the reason. It is something that should not happen and I'm reluctant that the API loses it's simplicity for developers by the need to understand why this context needs to get passed everywhere.

From my gut feeling, this can be all traced back to the Report Digester. I would rather see that one go instead of adding code to workaround that weirdness – it will haunt us forever if we do not tackle that properly now.

[the remark this answers is not in the document]
Comment (ramonski) [reply bika/lims/api.py]: Please could you prove that this only happens when you get the Analyses from the Digester. Try to disable the Digester completely and check if the error still occurs. If it does not, we will remove Digester in another PR

[the remark this answers is not in the document]
Comment (mikejmets) [reply bika/lims/api.py]: I merged the ARDigester methods in ARPublishView class and call _ar_data directly and the error does not occur.
